src/wp-includes/class-wp-block-scanner.php

```diff
@@ -0,0 +1,1244 @@
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $next_token_at = 0;
+
+	/**
+	 * Byte offset where namespace span begins.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $namespace_at = 0;
+
+	/**
+	 * Byte length of namespace span, or `0` if implicitly in the "core" namespace.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $namespace_length = 0;
+
+	/**
+	 * Byte offset where block name span begins.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $name_at = 0;
+
+	/**
+	 * Byte length of block name span.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $name_length = 0;
+
+	/**
+	 * Whether the delimiter contains the block-closing flag.
+	 *
+	 * This may be erroneous if present within a void block,
+	 * therefore the {@see self::has_closing_flag} can be used by
+	 * calling code to perform appropriate error-handling.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var bool
+	 */
+	private $has_closing_flag = false;
+
+	/**
+	 * Byte offset where JSON attributes span begins.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $json_at;
+
+	/**
+	 * Byte length of JSON attributes span, or `0` if none are present.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $json_length = 0;
+
+	/**
+	 * Internal parser state, differentiating whether the instance is currently matched,
+	 * on an implicit freeform node, in error, or ready to begin parsing.
+	 *
+	 * @see self::READY
+	 * @see self::MATCHED
+	 * @see self::HTML_OPEN
+	 * @see self::HTML_CLOSE
+	 * @see self::INCOMPLETE_INPUT
+	 * @see self::COMPLETE
+	 *
+	 * @since 6.9.0}
+	 *
+	 * @var string
+	 */
+	protected $state = self::READY;
+
+	/**
+	 * Indicates what kind of block comment delimiter this represents.
+	 *
+	 * One of:
+	 *
+	 *  - {@see static::OPENER} If the delimiter is opening a block.
+	 *  - {@see static::CLOSER} If the delimiter is closing an open block.
+	 *  - {@see static::VOID}   If the delimiter represents a void block with no inner content.
+	 *
+	 * If a parsed comment delimiter contains both the closing and the void
+	 * flags then it will be interpreted as a void block to match the behavior
+	 * of the official block parser, however, this is a mistake and probably
+	 * the block ought to close an open block of the same name, if one is open.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var string
+	 */
+	private $type;
+
+	/**
+	 * Creates a new block scanner.
+	 *
+	 * Example:
+	 *
+	 *     $scanner = WP_Block_Scanner::create( $html );
+	 *     while ( $scanner->next_delimiter() ) {
+	 *          if ( $scanner->opens_block( 'core/image' ) ) {
+	 *             echo "Found an image!\n";
+	 *         }
+	 *     }
+	 *
+	 * @see self::next_delimiter
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param string $source_text Input document potentially containing block content.
+	 * @return ?self Created block scanner, if successfully created.
+	 */
+	public static function create( string $source_text ) {
```

Comment:
From my comment on the commit https://github.com/WordPress/wordpress-develop/commit/85452871c61321a5c48f688fb3a94b16e051ca45#commitcomment-167116766

```suggestion
	public static function create( string $source_text ): WP_Block_Scanner {
```

Any subclass would naturally generate an object that is compatible with the type hint due to inheritance and abstraction rules, though in this case I think it's superfluous and confusing to hide the constructor then add a static `create` function, it's reinventing the constructor for no reason.

Reply:
Perhaps if this were a singleton it would make sense as a way to get the global object, but it isn't

Reply:
> Any subclass would naturally generate an object that is compatible with the type hint due to inheritance and abstraction rules

I’ve really struggled with adding static types here. If we use `self` or `?self` then the only option a subclass has is to return `?WP_Block_Scanner` and not, for instance, `?WP_Block_Processor` (or `?self`).

This is something that gets better with PHP 7.4.

<img width="999" height="529" alt="Screenshot 2025-10-03 at 11 00 20 AM" src="https://github.com/user-attachments/assets/863ce70d-6467-44fb-b650-1590ba88dffd" />

Even with `mixed` as a return type, it still crashes.

> it's superfluous and confusing to hide the constructor then add a static create function, it's reinventing the constructor for no reason.

The difference here is that a static method can return `null` but a constructor can only `throw` if it wants to reject the input arguments. I’m trying to leave room here to explore subclass usage before cutting off the possibility of creating a failable constructor.

At least with HTML it was important to be able to fail on the static creator methods as certain arrangements of arguments cannot be created.

In this case though I think you are right that the scanner class doesn’t need it at all. I’ll remove it. At some point I was hoping to have more harmony between the lower-level and higher-level classes by reusing the static constructor function. People can learn to do it two different ways though.

Reply:
I can foresee an empty string and a string of invalid blocks are the only failure cases, but not sure that this should result in a `null` as that implies an empty post is not possible

Reply:
this method is now gone and it’s impossible to fail when constructing a block processor.